### PR TITLE
Product-detail-default-options_CU-nner2e

### PR DESCRIPTION
**Pull request by a-wilkes**

### CODE REVIEW

**Comment by a-wilkes** [comment]
Task linked: [CU-nner2e Product-detail default options](https://app.clickup.com/t/nner2e)

**Comment by lpboucher** [comment]
Task linked: [CU-nner2e Product-detail default options](https://app.clickup.com/t/nner2e)

**Comment by lpboucher** [comment]
Task linked: [CU-nhbe2j Add quantity input to product](https://app.clickup.com/t/nhbe2j)

**Comment by a-wilkes** [comment]
I branched this from [Add-quantity-input-to-product](https://github.com/lpboucher/dak-coffee-wholesale/pull/74) since it already had the forms and controls set up which this PR needed to change/subscribe to changes to.

**Comment by lpboucher** [comment]
Got it. That's fine.

We should probably also merge develop into this branch and then change the destination branch to develop given that I merged a lot of tickets this weekend

**Review by lpboucher — CHANGES_REQUESTED** [review]
I think this is quite extensive and can likely be simplified.

I will update the ticket with more information. Maybe as an indication, we will use the directive (from the tutorial) that we had looked at previously

File: src/app/shared/directives/bind-query-params.directive.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    private formEqualsParams(formValues: any, paramValues: any): boolean {
+        const formEntries = Object.keys(formValues);
+        const paramEntries = Object.keys(paramValues);
+
+        return formEntries.length == paramEntries.length
+            && formEntries.reduce((res: boolean, key: string) => {
+                return formValues[key] == paramValues[key] && res;
+            }, true);
+    }
```
Comment (lpboucher) [inline src/app/shared/directives/bind-query-params.directive.ts]: I find this method a bit clunky. reduce is not correct operator here. Here are two solutions:

1) .every: use .every on the array returned from Object.keys and compare the values against the other object's values. If it is true one way it is also true the other way. .every returns a boolean so it fits nicely here
2) When trying to check deep equality of objects, you can use toJSON and compare the resulting string. The caveat is that property order is important so if we thing that order of properties is not guaranteed then we would have to sort the keys

Also, it is likely that this type of equality check will be required at some later point. I have moved the logic to a utils/helper file

**Review by lpboucher — APPROVED** [review]

**Pull request merged**